react_app/src/threeVisualizer.js

```diff
@@ -0,0 +1,230 @@
+
+      // Add the first face (the lower-right triangle)
+      var faceOne = new three.Face3(
+        this.vertices.length - 4,
+        this.vertices.length - 3,
+        this.vertices.length - 2
+      );
+
+      // Add the second face (the upper-left triangle)
+      var faceTwo = new three.Face3(
+        this.vertices.length - 4,
+        this.vertices.length - 2,
+        this.vertices.length - 1
+      );
+
+      // Add those faces to the geometry
+      this.faces.push(faceOne, faceTwo);
+
+      // Identify this subimage's offset in the x dimension
+      // An xOffset of 0 means the subimage starts flush with
+      // the left-hand edge of the atlas
+      var xOffset = i * (atlas.image.width / atlas.width);
+
+      // Identify the subimage's offset in the y dimension
+      // A yOffset of 0 means the subimage starts flush with
+      // the top edge of the atlas
+      var yOffset = 0;
+
+      // Use the xOffset and yOffset (and the knowledge that
+      // each row and column contains only 10 images) to specify
+      // the regions of the current image
+      this.faceVertexUvs[0].push([
+        new three.Vector2(xOffset, yOffset),
+        new three.Vector2(xOffset + (1.0 / atlas.numImages), yOffset),
+        new three.Vector2(xOffset + (1.0 / atlas.numImages), yOffset + 1.0)
+      ]);
+
+      // Map the region of the image described by the lower-left,
+      // upper-right, and upper-left vertices to `faceTwo`
+      this.faceVertexUvs[0].push([
+        new three.Vector2(xOffset, yOffset),
+        new three.Vector2(xOffset + (1.0 / atlas.numImages), yOffset + 1.0),
+        new three.Vector2(xOffset, yOffset + 1.0)
+      ]);
+    }
+  }
+}
+
+export class ThreeRenderer extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.dataset = props.dataset;
+    
+    this.canvasWidth = props.canvasWidth;
+    this.canvasHeight = props.canvasHeight;
+
+    this.camera = undefined;
+    this.scene = undefined;
+    this.controls = undefined;
+    this.material = undefined;
+    this.renderer = undefined;
+    this.canvas = undefined;
+    this.light = undefined;
+  }
+
+  // Loads json object for dataset and returns array of points
+  async initPoints(material) {
+    var numImages = 0;
+    const numRows = 1;
+
+    const json_url = '/coordinates-retrieval?dataset='.concat(this.dataset);
+    const result = await fetch(json_url);
+    const out = await result.json();
+
+    const points = out['points'].map((point_json) => {
+      const {x, y, z} = point_json;
+      return new Point(x, y, z);
+    });
+    
+    numImages = points.length;
+    const numColumns = numImages;
+
+    const atlas = new Atlas(new ImageParams(IMAGE_SIZE), numImages, numRows, numColumns);
+
+    const geometry = new AtlasGeometry(atlas, points);
+
+    const mesh = new three.Mesh(geometry, this.material);
+    mesh.position.set(0, 0, 0);
+    this.scene.add(mesh);
+
+    const component = this;
+    function render3d() {
```

Comment:
Why do you need to create a nested function for this?

Reply:
In this case it worked the best, @askewc and I worked on this for a bit and it seems to introduce rendering issues if it's not a nested function

Reply:
Oh.. because of the "requestAnimationFrame" which calls the self. That's fine with me (although there should be a better way of handling it; such as taking it out as an external function and adding a proper binding). Can you at least add a comment?

Reply:
Factored into member function with binding